### PR TITLE
feat: implement registry system (alias: loader)

### LINKED ISSUE #10: 🌱 functions loading: create a complete solution to load specific function
I'd like to suggest a feature that would allow for a more efficient use of memory within our projects. The idea is to develop a comprehensive solution for loading functions selectively through Sprout. This would enable developers to load either a single function or all functions, depending on their current needs, thereby optimizing the memory footprint.

Implementing this feature could significantly enhance performance and resource management, especially in environments where memory conservation is crucial.

I am eager to hear your thoughts and feedback on this proposal.

### PR DESCRIPTION
**Pull request by 42atomys**
## Description
Following the discussion about function loading in #31, this huge refactor implement each method behind multiples registries. 

> [!WARNING]
> This pull request are in progress, somes tests must be done before allowing it to be merge and release. Each feedback are welcome !

## TODO
- [x] Allow registry to register aliases with function
- [x] Test new package
- [x] Provide detailed documentation for each module and glossary for handler and registry (WIP but available [here](https://docs.atom.codes/sprout/~/changes/gP9HOVz2PoFWbplTHwVo))

## Changes
- Create the Handler and Registry interface
- Define the rules of a registry
- Add a Readme about how to create a registry
- Migrate all existing functions into separated registry
- Backward compatibility function `FuncsMap` register all registry by default 

## Checklist
- [x] I have read the **CONTRIBUTING.md** document.
- [x] My code follows the code style of this project.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] I have updated the documentation accordingly.
- [x] This change requires a change to the documentation on the website.

## Additional Information
This update following the discussions #31

### CODE REVIEW

**Comment by 42atomys** [comment]
First self feedback: allow registry to register aliases with functions

**Comment by 42atomys** [comment]
Second step : 
- test new package
- provide detailled documentation for each module and glossary for handler and registry

**Comment by 42atomys** [comment]
**Currently start the last step :** 
- Write a detailed documentation for each registry, a getting started and explanations for handler and registry, this will be a huge work to import it on Gitbook.

**🎉 This pull request are done in case of functionalities and tests. This will be merged soon as the documentation are ready too**

File: handler_test.go [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+	dh := &DefaultHandler{
+		cachedFuncsMap:   make(FunctionMap),
+		cachedFuncsAlias: make(FunctionAliasMap),
+	}
+
+	err := dh.AddRegistry(mockRegistry)
+	assert.NoError(t, err, "AddRegistry should not return an error")
+
+	assert.Len(t, dh.registries, 1, "Registry should be added to the DefaultHandler")
+	assert.Equal(t, mockRegistry, dh.registries[0], "Registry should match the mock registry")
```
Comment (ccoVeille) [inline handler_test.go:194]: I would use require.Len, otherwise it would report an error if empty, but then next line will panic


```suggestion
	require.Len(t, dh.registries, 1, "Registry should be added to the DefaultHandler")
	assert.Equal(t, mockRegistry, dh.registries[0], "Registry should match the mock registry")
```

Another way is to use assert.Contains


```suggestion
	assert.Len(t, dh.registries, 1, "Registry should be added to the DefaultHandler")
	assert.Contains(t, dh.registries, mockRegistry, "Registry should match the mock registry")
```

File: handler_test.go [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+	dh := &DefaultHandler{
+		cachedFuncsMap: make(FunctionMap),
+	}
+
+	err := dh.AddRegistries(mockRegistry1, mockRegistry2)
+	assert.NoError(t, err, "AddRegistries should not return an error")
+
+	assert.Len(t, dh.registries, 2, "Both registries should be added to the DefaultHandler")
+	assert.Equal(t, mockRegistry1, dh.registries[0], "First registry should match mockRegistry1")
+	assert.Equal(t, mockRegistry2, dh.registries[1], "Second registry should match mockRegistry2")
```
Comment (ccoVeille) [inline handler_test.go:169]: See https://github.com/go-sprout/sprout/pull/46#discussion_r1713052978

File: handler_test.go [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+
+	dh := &DefaultHandler{
+		cachedFuncsMap: make(FunctionMap),
+	}
+
+	err := dh.AddRegistry(mockRegistry)
+	assert.NoError(t, err, "AddRegistry should not return an error")
+
+	assert.Len(t, dh.registries, 1, "Registry should be added to the DefaultHandler")
+	assert.Equal(t, mockRegistry, dh.registries[0], "Registry should match the mock registry")
```
Comment (ccoVeille) [inline handler_test.go:142]: See https://github.com/go-sprout/sprout/pull/46#discussion_r1713052978

File: internal/helpers/helpers_test.go [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+func TestToString(t *testing.T) {
+	assert.Equal(t, "string", ToString("string"))
+	assert.Equal(t, "5", ToString(5))
+	assert.Equal(t, "true", ToString(true))
+	assert.Equal(t, "bytes", ToString([]byte("bytes")))
+	assert.Equal(t, "error", ToString(errors.New("error")))
+	assert.Equal(t, "stringer", ToString(testStringer{}))
+}
+
+func TestEmpty(t *testing.T) {
```
Comment (ccoVeille) [inline internal/helpers/helpers_test.go:25]: While this test is testing what is empty, it should also test that we get false for 1, .1, true, `" "`, []string{""}

File: internal/helpers/helpers_test.go [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+	// Testing for other types
+	assert.True(t, Empty([]byte{}))
+	assert.False(t, Empty(func() {}))
+}
+
+func TestUntilStep(t *testing.T) {
+	assert.Equal(t, []int{0, 2, 4, 6, 8}, UntilStep(0, 10, 2), "UntilStep: positive step")
+	assert.Equal(t, []int{10, 8, 6, 4, 2}, UntilStep(10, 0, -2), "UntilStep: negative step")
+	assert.Equal(t, []int{}, UntilStep(0, 10, 0), "UntilStep: invalid step")
+	assert.Equal(t, []int{}, UntilStep(10, 0, 2), "UntilStep: start > stop and positive step")
```
Comment (ccoVeille) [inline internal/helpers/helpers_test.go:60]: You could have use empty


```suggestion
	assert.Empty(t, UntilStep(0, 10, 0), "UntilStep: invalid step")
	assert.Empty(t, UntilStep(10, 0, 2), "UntilStep: start > stop and positive step")
```

File: internal/helpers/helpers_test.go [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+	tests := []struct {
+		name  string
+		input any
+		want  []string
+	}{
+		{"nil input", nil, []string{}},
+		{"[]string input", []string{"a", "b", "c"}, []string{"a", "b", "c"}},
+		{"[]any input", []any{"a", 1, true}, []string{"a", "1", "true"}},
+		{"[]int input", []int{1, 2, 3}, []string{"1", "2", "3"}},
+		{"[]float64 input", []float64{1.2, 3.4, 5.6}, []string{"1.2", "3.4", "5.6"}},
```
Comment (ccoVeille) [inline internal/helpers/helpers_test.go:73]: Please add edge cases life `7.` and `.8`

File: pesticide/test_helpers.go [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+
+	for _, test := range tc {
+		t.Run(test.Name, func(t *testing.T) {
+			t.Helper()
+
+			tmplResponse, err := runTemplate(t, testHandler(registry), test.Input, test.Data)
+			if test.ExpectedErr != "" {
+				if assert.Error(t, err) {
+					assert.Contains(t, err.Error(), test.ExpectedErr)
+				}
```
Comment (ccoVeille) [inline pesticide/test_helpers.go:59]: This can be replaced by


```suggestion
				assert.ErrorContains(t, err, test.ExpectedErr)
```

File: pesticide/test_helpers.go [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+}
+
+func runTemplate(t *testing.T, handler sprout.Handler, tmplString string, data any) (string, error) {
+	t.Helper()
+
+	tmpl, err := template.New("test").Funcs(handler.Build()).Parse(tmplString)
+	if err != nil {
+		assert.FailNow(t, "Failed to parse template", err)
+		return "", err
+	}
```
Comment (ccoVeille) [inline pesticide/test_helpers.go:81]:
```suggestion
	require.NoError(t, err, "Failed to parse template"
```

File: registry/checksum/functions.go [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+//
+// Returns:
+// - the Adler-32 checksum of the input string as a hexadecimal encoded string.
+//
+// Example:
+//
+// {{ adler32sum "Hello, World!" }} // Output: 1f9e046a
+func (cr *ChecksumRegistry) Adler32sum(input string) string {
+	hash := adler32.Checksum([]byte(input))
+	return fmt.Sprintf("%d", hash)
```
Comment (ccoVeille) [inline registry/checksum/functions.go:60]:
```suggestion
	return fmt.Sprint(hash)
```

File: registry/checksum/functions.go [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+// Parameters:
+// - input: the string to be hashed.
+//
+// Returns:
+// - the MD5 hash of the input string as a hexadecimal encoded string.
+//
+// Example:
+//
+// {{ md5sum "Hello, World!" }} // Output: 65a8e27d8879283831b664bd8b7f0ad4
+func (cr *ChecksumRegistry) Md5sum(input string) string {
```
Comment (ccoVeille) [inline registry/checksum/functions.go:75]: Wording suggestions

```suggestion
// SHA1sum calculates the SHA-1 hash of the input string and returns it as a
// hexadecimal encoded string.
//
// Parameters:
// - input: the string to be hashed.
//
// Returns:
// - the SHA-1 hash of the input string as a hexadecimal encoded string.
//
// Example:
//
// {{ sha1sum "Hello, World!" }} // Output: 0a0a9f2a6772942557ab5355d76af442f8f65e01
func (cr *ChecksumRegistry) SHA1sum(input string) string {
	hash := sha1.Sum([]byte(input))
	return hex.EncodeToString(hash[:])
}

// SHA256Sum calculates the SHA-256 hash of the input string and returns it as a
// hexadecimal encoded string.
//
// Parameters:
// - input: the string to be hashed.
//
// Returns:
// - the SHA-256 hash of the input string as a hexadecimal encoded string.
//
// Example:
//
// {{ sha256sum "Hello, World!" }} // Output: dffd6021bb2bd5b0af676290809ec3a53191dd81c7f70a4b28688a362182986f
func (cr *ChecksumRegistry) SHA256Sum(input string) string {
	hash := sha256.Sum256([]byte(input))
	return hex.EncodeToString(hash[:])
}

// Adler32Sum calculates the Adler-32 checksum of the input string and returns
// it as a hexadecimal encoded string.
//
// Parameters:
// - input: the string to be hashed.
//
// Returns:
// - the Adler-32 checksum of the input string as a hexadecimal encoded string.
//
// Example:
//
// {{ adler32sum "Hello, World!" }} // Output: 1f9e046a
func (cr *ChecksumRegistry) Adler32Sum(input string) string {
	hash := adler32.Checksum([]byte(input))
	return fmt.Sprintf("%d", hash)
}

// MD5Sum calculates the MD5 hash of the input string and returns it as a
// hexadecimal encoded string.
//
// Parameters:
// - input: the string to be hashed.
//
// Returns:
// - the MD5 hash of the input string as a hexadecimal encoded string.
//
// Example:
//
// {{ md5sum "Hello, World!" }} // Output: 65a8e27d8879283831b664bd8b7f0ad4
func (cr *ChecksumRegistry) MD5Sum(input string) string {
```

File: registry/crypto/functions.go [hunk 11]
```diff
@@ -0,0 +1,449 @@
...
+// input - the string to be hashed.
+// Returns the bcrypt hash as a string.
+//
+// Example:
+//
+//	{{ "Hello World" | bcrypt }} // Output: "$2a$12$C1qL8XVjIuGKzQXwC6g6tO"
+func (ch *CryptoRegistry) Bcrypt(input string) string {
+	hash, err := bcrypt_lib.GenerateFromPassword([]byte(input), bcrypt_lib.DefaultCost)
+	if err != nil {
+		return fmt.Sprintf("failed to encrypt string with bcrypt: %s", err)
```
Comment (ccoVeille) [inline registry/crypto/functions.go:39]: Here and everywhere

Why not wrapping the error?


```suggestion
		return fmt.Sprintf("failed to encrypt string with bcrypt: %w", err)
```

File: registry/crypto/functions.go [hunk 12]
```diff
@@ -0,0 +1,449 @@
...
+//	{{ generateSelfSignedCertificate "example.com" ["127.0.0.1"] ["localhost"] 365 }} // Output: {"Cert":"b64cert","Key":"b64key"}
+func (ch *CryptoRegistry) GenerateSelfSignedCertificate(
+	cn string,
+	ips []interface{},
+	alternateDNS []interface{},
+	daysValid int,
+) (Certificate, error) {
+	priv, err := rsa.GenerateKey(cryptorand.Reader, 2048)
+	if err != nil {
+		return Certificate{}, fmt.Errorf("error generating rsa key: %s", err)
```
Comment (ccoVeille) [inline registry/crypto/functions.go:270]:
```suggestion
		return Certificate{}, fmt.Errorf("error generating RSA key: %s", err)
```

File: registry/crypto/helpers.go [hunk 13]
```diff
@@ -0,0 +1,309 @@
...
+			x509.ExtKeyUsageClientAuth,
+		},
+		BasicConstraintsValid: true,
+	}, nil
+}
+
+func (ch *CryptoRegistry) pemBlockForKey(priv interface{}) *pem.Block {
+	switch k := priv.(type) {
+	case *rsa.PrivateKey:
+		return &pem.Block{Type: "RSA PRIVATE KEY", Bytes: x509.MarshalPKCS1PrivateKey(k)}
```
Comment (ccoVeille) [inline registry/crypto/helpers.go:104]: This type should be a constant, here and everywhere

Especially when there are code checking if type equal what is expected, it would help if there is a later refactoring

File: registry/crypto/helpers.go [hunk 14]
```diff
@@ -0,0 +1,309 @@
...
+	}
+}
+
+func (ch *CryptoRegistry) parsePrivateKeyPEM(pemBlock string) (crypto.PrivateKey, error) {
+	block, _ := pem.Decode([]byte(pemBlock))
+	if block == nil {
+		return nil, errors.New("no PEM data in input")
+	}
+
+	if block.Type == "PRIVATE KEY" {
```
Comment (ccoVeille) [inline registry/crypto/helpers.go:131]: Here is example


"PRIVATE KEY" should be a constant

File: registry/crypto/helpers.go [hunk 15]
```diff
@@ -0,0 +1,309 @@
...
+		priv, err := x509.ParsePKCS8PrivateKey(block.Bytes)
+		if err != nil {
+			return nil, fmt.Errorf("decoding PEM as PKCS#8: %s", err)
+		}
+		return priv, nil
+	} else if !strings.HasSuffix(block.Type, " PRIVATE KEY") {
```
Comment (ccoVeille) [inline registry/crypto/helpers.go:137]: This seems a bad way to do

It relies on the content of existing variable that are currently ok

I would prefer a switch with exact match of the 3 values that can validate this condition.

Because otherwise if someone wants to refactor or rename the EC PRIVATE KEY, they will never find or think about this line, and they will break things

**Comment by 42atomys** [comment]
Hi @ccoVeille, thanks for your contribution in reviewing this big PR 💜 

According to multiple comments on the crypto package, this package was backported from [sprig](https://github.com/Masterminds/sprig) and will be dropped in a few versions due to significant security issues.

I've been focusing on rewriting and enhancing other registries, so I haven't made any changes to this one.

File: registry/crypto/functions.go (hunk 12 above)
Comment (42atomys) [reply registry/crypto/functions.go:270]: Due to backward compatibility with Sprig and the lack of an accessible error variable to use with errors.As or errors.Is, most Sprig users who rely on the crypto package end up comparing strings. So, unfortunately, I cannot change it 😢

Also, please consider this discussion: https://github.com/go-sprout/sprout/pull/46#issuecomment-2282901894

File: registry/crypto/helpers.go (hunk 13 above)
Comment (42atomys) [reply registry/crypto/helpers.go:104]: Applicable to my comment: https://github.com/go-sprout/sprout/pull/46#issuecomment-2282901894

File: registry/crypto/helpers.go (hunk 14 above)
Comment (42atomys) [reply registry/crypto/helpers.go:131]: Applicable to my comment: https://github.com/go-sprout/sprout/pull/46#issuecomment-2282901894

File: registry/crypto/helpers.go (hunk 15 above)
Comment (42atomys) [reply registry/crypto/helpers.go:137]: Applicable to my comment: https://github.com/go-sprout/sprout/pull/46#issuecomment-2282901894

File: registry/crypto/functions.go (hunk 11 above)
Comment (42atomys) [reply registry/crypto/functions.go:39]: The first reason is that "the crypto registry hasn't changed; it's the same as in Sprig currently."

The second reason is that fmt.Sprintf does not support the error-wrapping directive %w, and many functions don't properly return errors. This is one of the main reasons why I want to evolve Sprig.

File: registry/checksum/functions.go (hunk 10 above)
Comment (42atomys) [reply registry/checksum/functions.go:75]: When I write the documentation this afternoon, I think about this change so I will do it ✔️

File: registry/crypto/functions.go (hunk 11 above)
Comment (ccoVeille) [reply registry/crypto/functions.go:39]: Oh I miss read and thought the signature was an error, so I thought it was an fmt.Errorf

**Comment by ccoVeille** [comment]
Old code is not funny. Noted

File: registry/crypto/functions.go (hunk 11 above)
Comment (42atomys) [reply registry/crypto/functions.go:39]: Don't worry the first time I try to touch that, I do the same mistake 👬

**Comment by caarlos0** [comment]
it looks really good! I liked the registry idea, and the implementation looks good!

lgtm!

**Comment by 42atomys** [comment]
Thanks @ccoVeille and @caarlos0 for your time and your contribution to reviewing this next step of Sprout 🌱 💜
